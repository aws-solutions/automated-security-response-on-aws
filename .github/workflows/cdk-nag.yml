name: CDK Nag Test

on:
  push:
  pull_request:
    types: [opened, edited, reopened, synchronize]

jobs:
  cdk-nag:
    name: CDK Nag Check
    if: ${{ (github.repository_owner != 'aws-solutions' && github.event_name == 'push') || (github.repository_owner == 'aws-solutions' && github.event_name == 'pull_request') }}
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [16.x]
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.node-version }}
      - run: |
          cd source
          npm ci
          cd solution_deploy
          npx cdk synth
          cd ../playbooks/AFSBP
          npx cdk synth
          cd ../CIS120
          npx cdk synth
          cd ../CIS140
          npx cdk synth
          cd ../NEWPLAYBOOK
          npx cdk synth
          cd ../PCI321
          npx cdk synth
          cd ../SC
          npx cdk synth
