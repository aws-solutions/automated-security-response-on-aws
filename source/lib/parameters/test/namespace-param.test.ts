// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { App, Aspects, DefaultStackSynthesizer, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AwsSolutionsChecks } from 'cdk-nag';
import NamespaceParam, { NAMESPACE_REGEX } from '../namespace-param';

function createNamespaceParamStack(): Stack {
  const app = new App();
  const stack = new Stack(app, 'NamespaceStack', {
    analyticsReporting: false,
    synthesizer: new DefaultStackSynthesizer({ generateBootstrapVersionRule: false }),
  });
  Aspects.of(stack).add(new AwsSolutionsChecks({ verbose: true }));
  new NamespaceParam(stack, 'NamespaceParam');
  return stack;
}

describe('namespace param stack', function () {
  const template = Template.fromStack(createNamespaceParamStack());

  it('matches snapshot', function () {
    expect(template).toMatchSnapshot();
  });

  describe('template param', function () {
    it('is present', function () {
      template.hasParameter('Namespace', {
        AllowedPattern: NAMESPACE_REGEX.source,
        Type: 'String',
      });
    });
  });
});

describe('namespace regex', function () {
  it('matches valid strings', function () {
    expect('my-test-n').toMatch(NAMESPACE_REGEX);
    expect('namespace').toMatch(NAMESPACE_REGEX);
    expect('123456789').toMatch(NAMESPACE_REGEX);
    expect('my2-e0lv').toMatch(NAMESPACE_REGEX);
  });

  it('does not match invalid strings', function () {
    expect('-test').not.toMatch(NAMESPACE_REGEX);
    expect('Namespace').not.toMatch(NAMESPACE_REGEX);
    expect('1234567890').not.toMatch(NAMESPACE_REGEX);
    expect('mY2-e0Lv').not.toMatch(NAMESPACE_REGEX);
    expect('nw').not.toMatch(NAMESPACE_REGEX);
    expect('sthree-nw').not.toMatch(NAMESPACE_REGEX);
    expect('xn--n').not.toMatch(NAMESPACE_REGEX);
    expect('nw-s3alias').not.toMatch(NAMESPACE_REGEX);
  });
});
