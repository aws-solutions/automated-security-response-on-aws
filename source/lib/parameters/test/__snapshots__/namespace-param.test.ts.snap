// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`namespace param stack matches snapshot 1`] = `
{
  "Parameters": {
    "Namespace": {
      "AllowedPattern": "(?!(^xn--|^sthree-|^sthree-configurator|^amzn-s3-demo-|.+-s3alias|.+--ol-s3|.+.mrap|.+--x-s3$))^[a-z0-9][a-z0-9-]{1,7}[a-z0-9]$",
      "ConstraintDescription": "The Namespace parameter must follow naming restrictions for S3 buckets and have a minimum length of 3 and a maximum length of 9. https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html",
      "Description": "Choose a unique namespace to be added as a suffix to remediation IAM role names. The same namespace should be used in the Member Roles and Member stacks. This string should be unique for each solution deployment.",
      "MaxLength": 9,
      "MinLength": 3,
      "Type": "String",
    },
  },
}
`;
